File: hospital-backend/src/main/java/com/hospital/backend/service/PricingService.java
package com.hospital.backend.service;

import com.hospital.backend.entity.PriceListType;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

/**
 * Service de tarification pour le calcul automatique des montants
 * lors des admissions et consultations.
 */
public interface PricingService {

    /**
     * Récupère le prix actuel d'un type de service
     * @param type Le type de service (FICHE, CONSULTATION_GENERALE, etc.)
     * @return Le prix unitaire ou null si non trouvé
     */
    BigDecimal getCurrentPrice(PriceListType type);

    /**
     * Récupère tous les prix actifs de la grille tarifaire
     * @return Liste des prix actifs
     */
    List<Map<String, Object>> getAllActivePrices();

    /**
     * Calcule le montant total pour une admission :
     * - Fiche (si nouveau patient)
     * - Consultation (selon le service choisi)
     * 
     * @param patientId ID du patient
     * @param serviceId ID du service médical
     * @return Map contenant les détails du calcul (ficheAmount, consulAmount, total)
     */
    Map<String, Object> calculateAdmissionAmount(Long patientId, Long serviceId);

    /**
     * Vérifie si le patient a déjà payé la fiche (validité 12 mois)
     * et retourne le montant de la fiche à payer (0 si déjà payée)
     * 
     * @param patientId ID du patient
     * @return Le montant de la fiche à payer (0 si exonéré)
     */
    BigDecimal getFicheAmount(Long patientId);

    /**
     * Vérifie si le patient possède une fiche active (payée dans les 12 derniers mois)
     * @param patientId ID du patient
     * @return true si le patient a une fiche active
     */
    boolean hasActiveFile(Long patientId);

    /**
     * Récupère la date de dernière paiement de la fiche
     * @param patientId ID du patient
     * @return La date du dernier paiement ou null
     */
    LocalDateTime getLastFichePaymentDate(Long patientId);

    /**
     * Met à jour un prix dans la grille tarifaire
     * @param id ID du prix à modifier
     * @param newPrice Nouveau prix
     * @return Le prix mis à jour
     */
    Map<String, Object> updatePrice(Long id, BigDecimal newPrice);

    /**
     * Crée un nouveau prix dans la grille
     * @param priceData Données du prix
     * @return Le prix créé
     */
    Map<String, Object> createPrice(Map<String, Object> priceData);

    /**
     * Initialise la grille tarifaire avec les valeurs par défaut
     * (Appelé au démarrage de l'application si la grille est vide)
     */
    void initializeDefaultPrices();
}

File: hospi-frontend/src/src/services/pricingService.js
import api from '@/services/api';

/**
 * Service API pour la gestion des tarifs et de la facturation
 */

// 1. Récupère tous les prix actifs de la grille tarifaire
export const getAllPrices = async () => {
  try {
    const response = await api.get('/pricing/prices');
    return response.data;
  } catch (error) {
    console.error('Erreur lors de la récupération des prix:', error);
    throw error;
  }
};

// 2. Calcule les montants pour une admission (NOUVELLE API CENTRALISÉE)
export const calculateAdmissionPrice = async (patientId, serviceId) => {
  try {
    const response = await api.get('/pricing/admission/calculate', {
      params: { patientId, serviceId }
    });
    return response.data;
  } catch (error) {
    console.error('Erreur lors du calcul des montants:', error);
    throw error;
  }
};

// 3. Vérifie le statut de la fiche patient
export const getPatientFicheStatus = async (patientId) => {
  try {
    const response = await api.get(`/pricing/patient/${patientId}/fiche-status`);
    return response.data;
  } catch (error) {
    console.error('Erreur lors de la vérification du statut fiche:', error);
    throw error;
  }
};

// 4. Crée un nouveau prix
export const createPrice = async (priceData) => {
  try {
    const response = await api.post('/pricing/prices', priceData);
    return response.data;
  } catch (error) {
    console.error('Erreur lors de la création du prix:', error);
    throw error;
  }
};

// 5. Met à jour un prix existant
export const updatePrice = async (id, priceData) => {
  try {
    const response = await api.put(`/pricing/prices/${id}`, priceData);
    return response.data;
  } catch (error) {
    console.error('Erreur lors de la mise à jour du prix:', error);
    throw error;
  }
};

// 6. Initialise la grille tarifaire avec les valeurs par défaut
export const initializePrices = async () => {
  try {
    const response = await api.post('/pricing/prices/initialize');
    return response.data;
  } catch (error) {
    console.error('Erreur lors de l\'initialisation des prix:', error);
    throw error;
  }
};

/**
 * EXPORT PAR DÉFAUT
 * Permet l'importation sous la forme : import pricingService from '...'
 */
const pricingService = {
  getAllPrices,
  calculateAdmissionPrice,
  getPatientFicheStatus,
  createPrice,
  updatePrice,
  initializePrices
};

export default pricingService;